src/Raven.Server/Https/ExternalCertificateValidator.cs

```diff
@@ -0,0 +1,288 @@
+
+            Task<CachedValue> task;
+            if (_externalCertificateValidationCallbackCache.TryGetValue(cacheKey, out var existingTask) == false)
+            {
+                task = new Task<CachedValue>(() => CheckExternalCertificateValidation(senderHostname, certificate, chain, sslPolicyErrors, log));
+                existingTask = _externalCertificateValidationCallbackCache.GetOrAdd(cacheKey, task);
+                if (existingTask == task)
+                {
+                    task.Start();
+
+                    if (_externalCertificateValidationCallbackCache.Count > 50)
+                    {
+                        foreach (var item in _externalCertificateValidationCallbackCache.Where(x => x.Value.IsCompleted).OrderBy(x => x.Value.Result.Until).Take(25))
+                        {
+                            _externalCertificateValidationCallbackCache.TryRemove(item.Key, out _);
+                        }
+                    }
+                }
+            }
+
+            CachedValue cachedValue;
+            try
+            {
+                cachedValue = existingTask.Result;
+            }
+            catch
+            {
+                _externalCertificateValidationCallbackCache.TryRemove(cacheKey, out _);
+                throw;
+            }
+
+            if (_server.Time.GetUtcNow() < cachedValue.Until)
+                return cachedValue.Valid;
+
+            var cachedValueNext = cachedValue.Next;
+            if (cachedValueNext != null)
+                return ReturnTaskValue(cachedValueNext);
+
+            task = new Task<CachedValue>(() =>
+                CheckExternalCertificateValidation(senderHostname, certificate, chain, sslPolicyErrors, log));
+
+            var nextTask = Interlocked.CompareExchange(ref cachedValue.Next, task, null);
+            if (nextTask != null)
+                return ReturnTaskValue(nextTask);
+
+
+            task.ContinueWith(done =>
+            {
+                _externalCertificateValidationCallbackCache.TryUpdate(cacheKey, done, existingTask);
+            }, TaskContinuationOptions.OnlyOnRanToCompletion);
+
+            task.Start();
+
+            return cachedValue.Valid; // we are computing this, but may take some time, let's use cached value for now
+
+            bool ReturnTaskValue(Task<CachedValue> task)
+            {
+                if (task.IsCompletedSuccessfully)
+                    return task.Result.Valid;
+
+                // not done yet? return the cached value
+                return cachedValue.Valid;
+            }
+        }
+
+        // https://stackoverflow.com/questions/14241479/how-to-split-a-space-delimited-list-of-paths-where-paths-can-include-spaces-in
+        private static readonly Regex re = new Regex(@"^([ ]*((?<r>[^ ""]+)|[""](?<r>[^""]*)[""]))*[ ]*$");
+
+        private static IEnumerable<string> RegexSplit(string input)
```

Comment:
Why do we need it now?

Reply:
We need to iterate over all the arguments (GetScriptLastWriteHash) and find their last write time because we don't know which argument is the actual script file. We do this so that if the script changes, we invalidate the cache value.

The special split function is used to split all the args even if they have a space in the path.
For example:
"Security.Certificate.Validation.Exec.Arguments": "-File \"c:\\work\\folder with spaces\\exec.ps1\" \"another argument\"",

Reply:
I do not like it. We should either expose an endpoint to clear this cache or just let user either wait few minutes or restart the server. This is a very rare situation. We should not do any magic around that.

Reply:
@ppekrol This is done so if you develop the script, we can give you a good experience of fixing issues as you test it.

Reply:
I understand the cause, but I'm afraid that this regex could not bulletproof. Especially if you will take into account all the platforms and all shells. It would be better IMO to add an endpoint to clear the cache.

Reply:
I would say that it would be better to add a method and be able to clear it from the JS Console, instead.